src/mlpack/core/data/one_hot_encoding_impl.hpp

```diff
@@ -56,14 +56,119 @@ void OneHotEncoding(const RowType& labelsIn,
     }
   }
   // Resize output matrix to necessary size, and fill it with zeros.
-  output.zeros(labelsIn.n_elem, curLabel);
+  output.zeros(curLabel, labelsIn.n_elem);
   // Fill ones in at the required places.
   for (size_t i = 0; i < labelsIn.n_elem; ++i)
   {
-    output(i, labels[i]) = 1;
+    output(labels[i], i) = 1;
   }
   labelMap.clear();
 }
+
+/**
+ * Overloaded function for the above function, which takes a matrix as input
+ * and also a vector of indices to encode and outputs a matrix.
+ * Indices represent the IDs of the dimensions to be one-hot encoded.
+ *
+ * @param input Input dataset to be encoded.
+ * @param indices Index of rows to be encoded.
+ * @param output Encoded matrix.
+ */
+template<typename eT>
+void OneHotEncoding(const arma::Mat<eT>& input,
+                    const arma::Col<size_t>& indices,
+                    arma::Mat<eT>& output)
+{
+  // Handle the edge case where there is nothing to encode.
+  if (indices.n_elem == 0)
+  {
+    output = input;
+    return;
+  }
+
+  // First, we need to compute the size of the output matrix.
+
+  // This vector will eventually hold the offsets for each dimension in the
+  // one-hot encoded matrix, but first it will just hold the counts of
+  // dimensions for each dimension.
+  arma::Col<size_t> dimensionOffsets(input.n_rows, arma::fill::ones);
+  // This will hold the mappings from a value that should be one-hot encoded to
+  // the index of the dimension it should take.
+  std::unordered_map<size_t, std::unordered_map<eT, size_t>> mappings;
+  for (size_t i = 0; i < indices.n_elem; ++i)
+  {
+    dimensionOffsets[indices[i]] = 0;
+    mappings.insert(
+        std::make_pair(indices[i], std::unordered_map<eT, size_t>()));
+  }
+
+  for (size_t col = 0; col < input.n_cols; ++col)
+  {
+    for (size_t row = 0; row < input.n_rows; ++row)
+    {
+      if (mappings.count(row) != 0)
+      {
+        // We have to one-hot encode this point.
+        if (mappings[row].count(input(row, col)) == 0)
+          mappings[row][input(row, col)] = dimensionOffsets[row]++;
+      }
+    }
+  }
+
+  // Turn the dimension counts into offsets.  Note that the last element is the
+  // total number of dimensions, and the first element is the offset for~
```

Comment:
Is the tilde at the end intentional?

Reply:
Oops, that's a bad paste from my editor (which highlights spaces at the end of lines as `~`).  We can remove it.